gdquest-godot-theme/addons/gdquest_docker/GDQuestDocker.gd

```diff
@@ -0,0 +1,131 @@
+tool
+extends EditorPlugin
+"""
+A container for plugins' widgets that provides an easy access to enable/disable valid plugins
+
+Notes
+---
+A valid plugin must have the EditorPlugin.get_plugin_name() -> String virtual method implemented
+and also a custom get_interface() -> Control method implemented.
+"""
+
+onready var interface: Docker
+
+const INTERFACE_SCENE: PackedScene = preload("res://addons/gdquest_docker/interface/Docker.tscn")
+const PLUGIN_CHECKBOX: PackedScene = preload("res://addons/gdquest_docker/interface/plugin_checkbox/PluginCheckBox.tscn")
+const ADDONS_FOLDER_PATH: String = "res://addons"
+
+
+func _enter_tree() -> void:
+	interface = INTERFACE_SCENE.instance()
+	add_control_to_dock(EditorPlugin.DOCK_SLOT_LEFT_UL, interface)
+	load_plugins()
+
+
+func _exit_tree() -> void:
+	var plugins: = interface.get_plugins_list()
+	var editor: = get_editor_interface()
+	
+	for checkbox in plugins.get_children():
+		if not checkbox.pressed:
+			continue
+		editor.set_plugin_enabled(checkbox.plugin_name, false)
+	
+	remove_control_from_docks(interface)
+
+
+func load_plugins() -> void:
+	var addons_folder: Directory = Directory.new()
+	if not addons_folder.dir_exists(ADDONS_FOLDER_PATH):
+		return
+	
+	addons_folder.open(ADDONS_FOLDER_PATH)
+	addons_folder.list_dir_begin(true)
+	var addon: String = addons_folder.get_next()
+	
+	while not addon == "":
+		add_plugin_to_list(addon)
+		addon = addons_folder.get_next()
+
+
+func add_plugin_to_list(plugin_name: String) -> void:
+	if plugin_name == get_plugin_name():
+		return
+	
+	var plugin_checkbox: PluginCheckBox = PLUGIN_CHECKBOX.instance()
+	plugin_checkbox.plugin_name = plugin_name
+	plugin_checkbox.connect("toggled", self, "set_plugin_enabled", [plugin_name])
+	interface.add_plugin_checkbox(plugin_checkbox)
+
+
+func remove_plugin_from_list(plugin_name: String) -> void:
+	var plugins_list: = interface.get_plugins_list()
+	
+	for c in plugins_list:
+		if c.plugin_name == plugin_name:
+			interface.remove_plugin_checkbox(c)
+			break
+
+
+func add_widget(widget : Control) -> void:
+	var container: = interface.get_widget_container()
+	container.add_child(widget)
+
+
+func remove_widget(widget : Control) -> void:
+	var container: = interface.get_widget_container()
+	container.remove_child(widget)
+
+
+func set_plugin_enabled(enabled: bool, plugin_name: String) -> void:
+	var editor: = get_editor_interface()
+	
+	if not editor.is_plugin_enabled(plugin_name):
+		if enabled:
+			editor.set_plugin_enabled(plugin_name, true)
+	
+	var plugin: = get_plugin(plugin_name)
```

Comment:
Do you have to get it every time or can you store the dependency once at the top of the script?

Reply:
I think is better to make it encapsulated here, and avoid a new mutable state. Since it can be any valid plugin, at any point in execution, and is only used in this method.

E.g.: we can have 5 valid plugins, if we store an `EditorPlugin` as a class dependency we will have a new mutation that will reference one of these for the whole execution, potentially losing reference to it, while we only need such reference here, when enabling/disabling them.

Reply:
I think is better to make it encapsulated here, and avoid a new mutable state. Since it can be any valid plugin, at any point in execution, and is only used in this method.

E.g.: we can have 5 valid plugins, if we store an `EditorPlugin` as a class dependency we will have a new mutation that will reference one of these for the whole execution, potentially losing reference to it, while we only need such reference here, when enabling/disabling them.

Reply:
That's fine. You could have a dict of plugins but alright, why not get it every time.